File: weight_tracker.js
#!/usr/bin/env node
// weight_tracker.js
const { program } = require('commander');
const fs = require('fs');
const chalk = require('chalk');

const DATA_FILE = 'weight.json';

class WeightTracker {
    constructor() {
        this.records = [];
        this.load();
    }

    load() {
        try {
            if (fs.existsSync(DATA_FILE)) {
                this.records = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
            }
        } catch (e) {
            this.records = [];
        }
        this.records.sort((a, b) => a.date.localeCompare(b.date));
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.records, null, 2));
    }

    addRecord(weight, date, photo) {
        if (!date) date = new Date().toISOString().split('T')[0];
        const record = { date, weight };
        if (photo) record.photo = photo;
        this.records.push(record);
        this.records.sort((a, b) => a.date.localeCompare(b.date));
        this.save();
        console.log(chalk.green(`Запись добавлена: ${date} - ${weight} кг${photo ? ` (фото: ${photo})` : ''}`));
    }

    listRecords() {
        if (this.records.length === 0) {
            console.log(chalk.yellow('Нет записей.'));
            return;
        }
        console.log(chalk.cyan('Дата       | Вес (кг) | Фото'));
        for (const r of this.records) {
            const photo = r.photo || '-';
            console.log(`${r.date} | ${r.weight.toFixed(1)} | ${photo}`);
        }
    }

    showChart() {
        if (this.records.length === 0) {
            console.log(chalk.yellow('Нет данных для графика.'));
            return;
        }
        const weights = this.records.map(r => r.weight);
        const minW = Math.min(...weights);
        const maxW = Math.max(...weights);
        if (maxW === minW) {
            console.log('Все значения одинаковы.');
            return;
        }
        const scale = 20 / (maxW - minW);
        console.log(chalk.cyan('График изменения веса (кг):'));
        for (const r of this.records) {
            const barLen = Math.floor((r.weight - minW) * scale) + 1;
            const bar = '█'.repeat(barLen);
            const photo = r.photo || '';
            console.log(`${r.date} ${chalk.green(bar)} ${r.weight.toFixed(1)}  ${photo}`);
        }
    }

    showStats() {
        if (this.records.length === 0) {
            console.log(chalk.yellow('Нет данных.'));
            return;
        }
        const weights = this.records.map(r => r.weight);
        const minW = Math.min(...weights);
        const maxW = Math.max(...weights);
        const avgW = weights.reduce((a, b) => a + b, 0) / weights.length;
        const first = weights[0];
        const last = weights[weights.length - 1];
        const diff = last - first;
        const trend = diff > 0 ? '📈 растёт' : diff < 0 ? '📉 падает' : '➡️ стабилен';
        console.log(chalk.cyan('Статистика:'));
        console.log(`  Минимальный: ${chalk.green(minW.toFixed(1))} кг`);
        console.log(`  Максимальный: ${chalk.red(maxW.toFixed(1))} кг`);
        console.log(`  Средний: ${chalk.yellow(avgW.toFixed(1))} кг`);
        console.log(`  Тренд: ${chalk.magenta(trend)} (${diff.toFixed(1)} кг)`);
    }

    export(filename) {
        const ext = filename.split('.').pop().toLowerCase();
        let content;
        if (ext === 'json') {
            content = JSON.stringify(this.records, null, 2);
        } else if (ext === 'csv') {
            const header = 'date,weight,photo\n';
            const rows = this.records.map(r => `${r.date},${r.weight},${r.photo || ''}`).join('\n');
            content = header + rows;
        } else {
            console.log(chalk.red('Неподдерживаемый формат. Используйте .json или .csv'));
            return;
        }
        fs.writeFileSync(filename, content);
        console.log(chalk.green(`Экспортировано в ${filename}`));
    }
}

program
    .option('--add <weight>', 'Добавить вес (кг)', parseFloat)
    .option('--date <date>', 'Дата (YYYY-MM-DD)')
    .option('--photo <path>', 'Путь к фотографии')
    .option('--list', 'Показать записи')
    .option('--chart', 'Показать график')
    .option('--stats', 'Показать статистику')
    .option('--export <file>', 'Экспорт в файл')
    .parse(process.argv);

const opts = program.opts();
const tracker = new WeightTracker();

if (opts.add !== undefined) {
    tracker.addRecord(opts.add, opts.date, opts.photo);
} else if (opts.list) {
    tracker.listRecords();
} else if (opts.chart) {
    tracker.showChart();
} else if (opts.stats) {
    tracker.showStats();
} else if (opts.export) {
    tracker.export(opts.export);
} else {
    program.help();
}

File: WeightTracker.cs
// WeightTracker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace WeightTracker
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            var tracker = new Tracker();
            if (opts.Add.HasValue)
            {
                tracker.AddRecord(opts.Add.Value, opts.Date, opts.Photo);
            }
            else if (opts.List)
            {
                tracker.ListRecords();
            }
            else if (opts.Chart)
            {
                tracker.ShowChart();
            }
            else if (opts.Stats)
            {
                tracker.ShowStats();
            }
            else if (opts.Export != null)
            {
                tracker.Export(opts.Export);
            }
            else
            {
                Console.WriteLine("Используйте --help для справки.");
            }
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--add": opts.Add = double.Parse(args[++i]); break;
                    case "--date": opts.Date = args[++i]; break;
                    case "--photo": opts.Photo = args[++i]; break;
                    case "--list": opts.List = true; break;
                    case "--chart": opts.Chart = true; break;
                    case "--stats": opts.Stats = true; break;
                    case "--export": opts.Export = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public double? Add { get; set; }
            public string Date { get; set; }
            public string Photo { get; set; }
            public bool List { get; set; }
            public bool Chart { get; set; }
            public bool Stats { get; set; }
            public string Export { get; set; }
        }

        class Record
        {
            public string Date { get; set; }
            public double Weight { get; set; }
            public string Photo { get; set; }
        }

        class Tracker
        {
            private const string DataFile = "weight.json";
            private List<Record> records = new List<Record>();

            public Tracker() => Load();

            private void Load()
            {
                try
                {
                    if (File.Exists(DataFile))
                    {
                        string json = File.ReadAllText(DataFile);
                        records = JsonSerializer.Deserialize<List<Record>>(json) ?? new List<Record>();
                    }
                }
                catch { records = new List<Record>(); }
                records = records.OrderBy(r => r.Date).ToList();
            }

            private void Save()
            {
                string json = JsonSerializer.Serialize(records, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(DataFile, json);
            }

            public void AddRecord(double weight, string date, string photo)
            {
                if (string.IsNullOrEmpty(date)) date = DateTime.UtcNow.ToString("yyyy-MM-dd");
                records.Add(new Record { Date = date, Weight = weight, Photo = photo });
                records = records.OrderBy(r => r.Date).ToList();
                Save();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.Write($"Запись добавлена: {date} - {weight} кг");
                if (!string.IsNullOrEmpty(photo)) Console.Write($" (фото: {photo})");
                Console.WriteLine();
                Console.ResetColor();
            }

            public void ListRecords()
            {
                if (records.Count == 0)
                {
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    Console.WriteLine("Нет записей.");
                    Console.ResetColor();
                    return;
                }
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("Дата       | Вес (кг) | Фото");
                Console.ResetColor();
                foreach (var r in records)
                {
                    string photo = !string.IsNullOrEmpty(r.Photo) ? r.Photo : "-";
                    Console.WriteLine($"{r.Date} | {r.Weight:F1} | {photo}");
                }
            }

            public void ShowChart()
            {
                if (records.Count == 0)
                {
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    Console.WriteLine("Нет данных для графика.");
                    Console.ResetColor();
                    return;
                }
                var weights = records.Select(r => r.Weight).ToList();
                double minW = weights.Min();
                double maxW = weights.Max();
                if (Math.Abs(maxW - minW) < 1e-9)
                {
                    Console.WriteLine("Все значения одинаковы.");
                    return;
                }
                double scale = 20.0 / (maxW - minW);
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("График изменения веса (кг):");
                Console.ResetColor();
                foreach (var r in records)
                {
                    int barLen = (int)((r.Weight - minW) * scale) + 1;
                    string bar = new string('█', barLen);
                    string photo = !string.IsNullOrEmpty(r.Photo) ? " " + r.Photo : "";
                    Console.ForegroundColor = ConsoleColor.Green;
                    Console.WriteLine($"{r.Date} {bar} {r.Weight:F1}{photo}");
                    Console.ResetColor();
                }
            }

            public void ShowStats()
            {
                if (records.Count == 0)
                {
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    Console.WriteLine("Нет данных.");
                    Console.ResetColor();
                    return;
                }
                var weights = records.Select(r => r.Weight).ToList();
                double minW = weights.Min();
                double maxW = weights.Max();
                double avgW = weights.Average();
                double first = weights.First();
                double last = weights.Last();
                double diff = last - first;
                string trend = diff > 0 ? "📈 растёт" : diff < 0 ? "📉 падает" : "➡️ стабилен";
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("Статистика:");
                Console.ResetColor();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"  Минимальный: {minW:F1} кг");
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine($"  Максимальный: {maxW:F1} кг");
                Console.ForegroundColor = ConsoleColor.Yellow;
                Console.WriteLine($"  Средний: {avgW:F1} кг");
                Console.ForegroundColor = ConsoleColor.Magenta;
                Console.WriteLine($"  Тренд: {trend} ({diff:+F1} кг)");
                Console.ResetColor();
            }

            public void Export(string filename)
            {
                string ext = Path.GetExtension(filename).ToLower().TrimStart('.');
                string content;
                if (ext == "json")
                {
                    content = JsonSerializer.Serialize(records, new JsonSerializerOptions { WriteIndented = true });
                }
                else if (ext == "csv")
                {
                    using var sw = new StringWriter();
                    using var csv = new CsvHelper.CsvWriter(sw, System.Globalization.CultureInfo.InvariantCulture);
                    csv.WriteRecords(records);
                    content = sw.ToString();
                }
                else
                {
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.WriteLine("Неподдерживаемый формат. Используйте .json или .csv");
                    Console.ResetColor();
                    return;
                }
                File.WriteAllText(filename, content);
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Экспортировано в {filename}");
                Console.ResetColor();
            }
        }
    }
}
